worker/rhpv3.go

```diff
@@ -498,111 +442,131 @@ func readSectorCost(pt *rhpv3.HostPriceTable) (types.Currency, error) {
+	priceTableUpdateLeeway := -time.Duration(frand.Intn(60)) * time.Second
+	if time.Now().Before(hpt.Expiry.Add(priceTableValidityLeeway).Add(priceTableUpdateLeeway)) {
+		return
 	}
-	pt.mu.Unlock()
 
-	// If this thread is not supposed to perform the update, just block and
-	// return the result.
-	if !performUpdate {
+	// price table is valid and update ongoing, return early
+	ongoing, update := p.ongoingUpdate()
+	if ongoing && time.Now().Before(hpt.Expiry.Add(priceTableValidityLeeway)) {
+		return
+	}
+
+	// price table is being updated, wait for the update
+	if ongoing {
 		select {
 		case <-ctx.Done():
-			return rhpv3.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
-		case <-ongoing.done:
-		}
-		if ongoing.err != nil {
-			return rhpv3.HostPriceTable{}, ongoing.err
-		} else {
-			return *ongoing.pt, nil
+			return hostdb.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
+		case <-update.done:
 		}
+		return update.hpt, update.err
 	}
 
-	// Update price table.
-	var hpt rhpv3.HostPriceTable
-	err := withTransportV3(ctx, siamuxAddr, hk, func(t *rhpv3.Transport) (err error) {
-		hpt, err = RPCPriceTable(t, payFn)
-		return err
-	})
+	// this thread is updating the price table
+	defer func() {
+		update.hpt = hpt
+		update.err = err
+		close(update.done)
 
-	pt.mu.Lock()
-	defer pt.mu.Unlock()
+		p.mu.Lock()
+		if err == nil {
+			p.hpt = hpt
+		}
+		p.update = nil
+		p.mu.Unlock()
+	}()
 
-	// On success we update the pt.
-	if err == nil {
-		ongoing.pt = &hpt
-		pt.pt = &hpt
-		pt.expiry = time.Now().Add(hpt.Validity)
+	// fetch the host info from the bus, it might have a new price table already
+	// in which case we can return without having to update it
+	host, err := b.Host(ctx, hk)
+	if err == nil && host.PriceTable != nil && time.Now().Before(host.PriceTable.Expiry.Add(priceTableValidityLeeway)) {
+		hpt = *host.PriceTable
+		return
+	}
+
+	// try to pay by EA if possible
+	if revision == nil {
+		if cs, err1 := b.ConsensusState(ctx); err1 == nil && cs.Synced {
+			if pt, err2 := w.fetchPriceTable(ctx, hk, host.Settings.SiamuxAddr(), w.preparePriceTableAccountPayment(hk, cs.BlockHeight)); err2 == nil {
+				hpt = pt
+				return
+			}
+		}
 	}
-	// Signal that the update is over.
-	ongoing.err = err
-	close(ongoing.done)
-	pt.ongoingUpdate = nil
-	return hpt, err
-}
 
-// priceTable returns a priceTable from priceTables for the given host or
-// creates a new one.
-func (pts *priceTables) priceTable(hk types.PublicKey) *priceTable {
-	pts.mu.Lock()
-	defer pts.mu.Unlock()
+	// pay by FC if EA payment failed or if a revision was given
+	if revision != nil {
+		return w.fetchPriceTable(ctx, hk, host.Settings.SiamuxAddr(), w.preparePriceTableContractPayment(hk, revision))
+	}
 
-	if _, exists := pts.priceTables[hk]; !exists {
-		pts.priceTables[hk] = &priceTable{hk: hk}
+	// fetch the active contract
+	contract, err := b.ActiveContract(ctx, hk)
```

Comment:
Is this disregarding the active contract set? So we are potentially spending from a contract that the autopilot doesn't consider good? To be fair I'd rather avoid that and fail if account payment doesn't work.

We pass a revision in 99% of cases anyway except for endpoints where we don't have access to one. Which right now is only downloads where we fail anyway if the account is empty.

Reply:
That's a good point, I didn't think of that. Removing the fallback to contract payments makes everything a lot simpler so I did that for the time being.